### PR TITLE
Add Error Prone and fix or suppress warnings.

### PR DESCRIPTION
**Pull request by NightlyNexus**

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -73,4 +75,12 @@ subprojects {
   afterEvaluate {
     tasks.findByName('check').dependsOn('checkstyle')
   }
-}
\ No newline at end of file
+
+  if (isCi) {
```
Comment (JakeWharton) [inline build.gradle]: Let's just always enable this. I only do conditional on projects that might be affected by the slower compiles.

File: build.gradle [hunk 2]
```diff
@@ -27,21 +28,22 @@ buildscript {
...
+    classpath 'net.ltgt.gradle:gradle-errorprone-plugin:0.0.13'
   }
 }
 
 apply plugin: 'com.github.ben-manes.versions'
 
 subprojects {
   repositories {
-    jcenter()
+    mavenCentral()
```
Comment (jrodbx) [inline build.gradle]: does the plugin require this?  I thought plugins only require buildscript dependencies.

Comment (jrodbx) [reply build.gradle]: if not, let's remove.  else move under `google()`.

File: build.gradle [hunk 3]
```diff
@@ -73,4 +75,10 @@ subprojects {
...
     tasks.findByName('check').dependsOn('checkstyle')
   }
-}
\ No newline at end of file
+
+  apply plugin: 'net.ltgt.errorprone'
+
+  dependencies {
+    errorprone deps.errorProne
+  }
```
Comment (jrodbx) [inline build.gradle]: I'm guessing this is related to the buildscript only comment.

File: picasso-provider/src/main/java/com/squareup/picasso3/provider/PicassoProvider.java [hunk 4]
```diff
@@ -22,7 +22,7 @@
  */
 public final class PicassoProvider {
   @SuppressLint("StaticFieldLeak")
-  private static Picasso instance;
+  private static volatile Picasso instance;
```
Comment (jrodbx) [inline picasso-provider/src/main/java/com/squareup/picasso3/provider/PicassoProvider.java]: hmm. would we ever want this to be accessed off the main thread?

File: picasso/src/main/java/com/squareup/picasso3/AssetRequestHandler.java [hunk 5]
```diff
@@ -32,7 +32,7 @@
 
   private final Context context;
   private final Object lock = new Object();
-  private AssetManager assetManager;
+  private volatile AssetManager assetManager;
```
Comment (jrodbx) [inline picasso/src/main/java/com/squareup/picasso3/AssetRequestHandler.java]: ditto

File: picasso/src/main/java/com/squareup/picasso3/FetchAction.java [hunk 6]
```diff
@@ -19,6 +19,7 @@
 
 class FetchAction extends Action<Object> {
 
+  @SuppressWarnings("HidingField")
```
Comment (jrodbx) [inline picasso/src/main/java/com/squareup/picasso3/FetchAction.java]: what's this do?

File: picasso/src/main/java/com/squareup/picasso3/Utils.java [hunk 7]
```diff
@@ -336,8 +337,7 @@ static void flushStackLocalLeaks(Looper looper) {
   }
 
   static class PicassoThreadFactory implements ThreadFactory {
-    @SuppressWarnings("NullableProblems")
-    public Thread newThread(Runnable r) {
+    @Override public Thread newThread(@NonNull Runnable r) {
```
Comment (jrodbx) [inline picasso/src/main/java/com/squareup/picasso3/Utils.java]: we've been removing NonNulls in favor of defaults, and only marking Nullable parameters.

File: picasso/src/test/java/com/squareup/picasso3/RequestCreatorTest.java [hunk 8]
```diff
@@ -180,7 +180,7 @@ public void intoTargetWithFitThrows() {
     }
   }
 
-  public void intoTargetNoPlaceholderCallsWithNull() {
+  @Test public void intoTargetNoPlaceholderCallsWithNull() {
```
Comment (jrodbx) [inline picasso/src/test/java/com/squareup/picasso3/RequestCreatorTest.java]: haha

File: picasso-provider/src/main/java/com/squareup/picasso3/provider/PicassoProvider.java (hunk 4 above)
Comment (NightlyNexus) [reply picasso-provider/src/main/java/com/squareup/picasso3/provider/PicassoProvider.java]: Probably not. Just matches what existed before.

File: picasso/src/main/java/com/squareup/picasso3/FetchAction.java (hunk 6 above)
Comment (NightlyNexus) [reply picasso/src/main/java/com/squareup/picasso3/FetchAction.java]: super.target is a different thing.

File: picasso/src/main/java/com/squareup/picasso3/Utils.java (hunk 7 above)
Comment (NightlyNexus) [reply picasso/src/main/java/com/squareup/picasso3/Utils.java]: IntelliJ gives a warning here cuz the supertype parameter is marked/known non-null.

File: build.gradle (hunk 2 above)
Comment (NightlyNexus) [reply build.gradle]: Unrelated change, yeah.

Comment (NightlyNexus) [reply build.gradle]: okay, looking on my computer now...
this change isn't related to the plugin. i just figured we prefer mc over jcenter.
we need one of them for the dependencies like OkHttp, etc.
`google()` is below.

what are you asking?

File: build.gradle (hunk 3 above)
Comment (NightlyNexus) [reply build.gradle]: i don't follow. what comment?

File: picasso-provider/src/main/java/com/squareup/picasso3/provider/PicassoProvider.java (hunk 4 above)
Comment (NightlyNexus) [reply picasso-provider/src/main/java/com/squareup/picasso3/provider/PicassoProvider.java]: to match #882

File: build.gradle (hunk 3 above)
Comment (jrodbx) [reply build.gradle]: ah, never mind, read the errorprone README: https://github.com/tbroyer/gradle-errorprone-plugin#usage

File: picasso/src/main/java/com/squareup/picasso3/FetchAction.java (hunk 6 above)
Comment (jrodbx) [reply picasso/src/main/java/com/squareup/picasso3/FetchAction.java]: I renamed this and the other field to resolve the hidden/scoping issue.

**Comment by jrodbx** [comment]
Merged in 3a883f052eca677a1c047c0d9e2b43a3bdaa1e4e

File: build.gradle (hunk 2 above)
Comment (jrodbx) [reply build.gradle]: understood.  when I originally reviewed this (on mobile), I thought this diff related to the subject line of this PR ("errorprone").

**Pull request closed without merging**